### DISCUSSION TITLE
Measuring codebase health

### DISCUSSION THREAD

**Discussion by sberrevoets** [MobileNativeFoundation/discussions#107] [General]
Last year we started an initiative at Lyft to measure the overall health of the codebase through predetermined metrics that we're planning on expanding over time. Currently we use it to look primarily at code quality: class/method lengths, number of parameters per method, cyclomatic complexity, that kind of stuff.

For each measure we define ranges for what we consider reasonable as far as code quality goes (e.g. a class with more than 500 lines is too long), and then highlight any issues with their severity (critical/error/warning). We break this down per module, and since every module at Lyft is owned by a team we can show one dimension of tech debt, broken down by module or by team. This gives teams the insight they need to prioritize and fix this tech debt.

Over time we're looking to add more information like crashes, open Jira tickets, outdated technologies, etc. to look at the quality of their features more holistically.

There are some third party services that do similar things, most notably [SonarQube](http://sonarqube.org) and [codebeat](codebeat.co). In #106 I saw Slack does similar things, and I'm curious if other companies have similar (or entirely different) metrics/data they look at and how they go about addressing any issues the tools surface.

**Comment by vRallev** [comment]
We (Square) are very interested in this area, too. We came to a very similar conclusion to have a dashboard for the entire codebase and separate views for individual teams. We didn't roll out a solution yet. We're still exploring which technology works best for us. [One such exploration](https://github.com/vRallev/compose-desktop-exploration) was during a hack week written in Compose.

In regards to metrics we show several migrations that need to be performed in this module (e.g. adopt [Anvil](https://github.com/square/anvil)), whether the module is too big / too small, a mixed Kotlin / Java module for Android, uses KAPT, number of disabled tests, lines of code and cumulative lines of code. We plan to add more in the future.

**Comment by fdiaz** [comment]
We (Airbnb) have a similar type of system on iOS that we call "Module Score report". We just launched this a couple of weeks ago, so it's hard to measure impact just yet.

Basically what we did is we assign a set of metrics that measure usage of "modern technology" (e.g. usage of our internal framework for propagating data, usage of DI, etc) and usage of "bad patterns" (dependencies on deprecated modules, usage of [non-inclusive terminology](https://link.medium.com/3Fqx0DA76gb), etc). With this information we calculate a score (currently a percentage but we're evaluating moving to a grade system) and assign it per-module and then per-team (similar to you each module has an owning team).

Technically to create these reports we rely heavily on [reading the AST](https://github.com/fdiaz/SwiftInspector) and using [Buck](https://github.com/airbnb/BuckSample) to query information about modules, so it's all built in-house.

We also have a concept called Test Levels that measure how well tested a module is. This has been rolled out for a longer period and is a company-wide initiative that has been quite successful. These levels are reported to leadership and as such some teams have defined OKRs around improving them.

**Comment by sgleadow** [comment]
We don't have a specific measure of mobile codebase health. I could imagine having an assessment that included some automated code checks, linting, complexity, dependencies and some operational reporting like crashes, errors and deployment frequency.

In terms of high level metrics, they're not mobile code specific, but the [DORA metrics](https://cloud.google.com/blog/products/devops-sre/using-the-four-keys-to-measure-your-devops-performance) are a useful and proven set of things to measure. It also forces us to think about better ways to achieve a good "Lead time for changes" and "Time to restore" in the world of app store releases.

At REA, we rate every system in the company every 3 months against a common systems health criteria. This gives us a high level view of system health (but not as specific as code complexity etc). Over time, we see trends in system health and teams have a 20% time allocation towards maintaining the health of the systems they own (for some definition of 20% 😄 ).

The criteria evolve over time as to what we think "good software" looks like, and I don't have public version of the latest criteria, but to get an idea, here's a [public tech blog from one of my colleagues on what it looked like in 2017](https://www.rea-group.com/blog/what-good-software-looks-like-at-rea/).

**Comment by ferologics** [comment]
For smaller teams that cannot afford to develop in-house code quality analysis solutions [CodeClimate](https://codeclimate.com) and [CodeCov](https://about.codecov.io) give you similar metrics about your code base. They're not as sophisticated but get the job done and get out of your way.
